ethereum/p2p/src/main/java/tech/pegasys/pantheon/ethereum/p2p/upnp/UpnpNatManager.java

```diff
@@ -0,0 +1,479 @@
+/*
+ * Copyright 2018 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+package tech.pegasys.pantheon.ethereum.p2p.upnp;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.jupnp.UpnpService;
+import org.jupnp.UpnpServiceImpl;
+import org.jupnp.model.action.ActionInvocation;
+import org.jupnp.model.message.UpnpResponse;
+import org.jupnp.model.message.header.STAllHeader;
+import org.jupnp.model.meta.RemoteDevice;
+import org.jupnp.model.meta.RemoteDeviceIdentity;
+import org.jupnp.model.meta.RemoteService;
+import org.jupnp.model.types.UnsignedIntegerFourBytes;
+import org.jupnp.model.types.UnsignedIntegerTwoBytes;
+import org.jupnp.registry.Registry;
+import org.jupnp.registry.RegistryListener;
+import org.jupnp.support.igd.callback.GetExternalIP;
+import org.jupnp.support.igd.callback.GetStatusInfo;
+import org.jupnp.support.igd.callback.PortMappingAdd;
+import org.jupnp.support.model.Connection;
+import org.jupnp.support.model.PortMapping;
+
+/**
+ * Manages underlying UPnP library "jupnp" and provides abstractions for asynchronously interacting
+ * with the NAT environment through UPnP.
+ *
+ * <p>This class is not thread-safe.
```

Comment:
What operations are not thread safe?  Can we syncornize those methods or add a lock object we lock on?

Reply:
I added some synchronization that should make it threadsafe, see e140b27.

Reply:
Question:  is the synchronization here just to protect the `recognizedServices` map and to respect the `started` status?